File: PandorasBox/Ahorn/entities/lasersensor.jl
module PandorasBoxLaserSensor

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/laserSensor" LaserSensor(x::Integer, y::Integer, flag::String="", color::String="White", blockLight::Bool=true, mode::String="Continuous")

const colors = sort(collect(keys(Ahorn.XNAColors.colors)))
const modes = String["Continuous", "HitOnce"]

const placements = Ahorn.PlacementDict(
    "Laser Sensor (Pandora's Box)" => Ahorn.EntityPlacement(
        LaserSensor,
        "point"
    )
)

Ahorn.editingOptions(entity::LaserSensor) = Dict{String, Any}(
    "color" => colors,
    "mode" => modes
)

function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

orbTexture = "objects/pandorasBox/laser/sensor/orb"
metalRingTexture = "objects/pandorasBox/laser/sensor/metal_ring"
lightRingTexture = "objects/pandorasBox/laser/sensor/light_ring"

function Ahorn.selection(entity::LaserSensor)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle(metalRingTexture, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LaserSensor, room::Maple.Room)
    colorRaw = entity.color
    color = getColor(colorRaw)

    Ahorn.drawSprite(ctx, metalRingTexture, 0, 0)
    Ahorn.drawSprite(ctx, lightRingTexture, 0, 0, tint=color)
    Ahorn.drawSprite(ctx, orbTexture, 0, 0)
end

end

File: PandorasBox/Loenn/entities/laser_sensor.lua
local drawableSprite = require("structs.drawable_sprite")

local lightSensor = {}

local modes = {
    "Continuous",
    "HitOnce"
}

lightSensor.name = "pandorasBox/laserSensor"
lightSensor.depth = 200
lightSensor.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}
lightSensor.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    },
    mode = {
        options = modes,
        editable = false
    }
}
lightSensor.placements = {
    name = "sensor",
    data = {
        flag = "",
        color = "White",
        blockLight = true,
        mode = "Continuous"
    }
}

local orbTexture = "objects/pandorasBox/laser/sensor/orb"
local metalRingTexture = "objects/pandorasBox/laser/sensor/metal_ring"
local lightRingTexture = "objects/pandorasBox/laser/sensor/light_ring"


function lightSensor.sprite(room, entity)
    local colorRaw = entity.color or "White"
    -- Don't want color to carry over
    local position = {
        x = entity.x,
        y = entity.y
    }

    local metalRingSprite = drawableSprite.fromTexture(metalRingTexture, position)
    local lightRingSprite = drawableSprite.fromTexture(lightRingTexture, position)
    local orbSprite = drawableSprite.fromTexture(orbTexture, position)

    lightRingSprite:setColor(colorRaw)

    return {
        metalRingSprite,
        lightRingSprite,
        orbSprite
    }
end

return lightSensor